src/PairedAmpRouting.php

```diff
@@ -0,0 +1,562 @@
+		$this->flush_rewrite_rules();
+	}
+
+	/**
+	 * Register.
+	 */
+	public function register() {
+		add_filter( 'amp_default_options', [ $this, 'filter_default_options' ], 10, 2 );
+		add_filter( 'amp_options_updating', [ $this, 'sanitize_options' ], 10, 2 );
+
+		add_action( 'init', [ $this, 'add_rewrite_endpoint' ], 0 );
+		add_action( 'parse_query', [ $this, 'correct_query_when_is_front_page' ] );
+
+		add_filter( 'old_slug_redirect_url', [ $this, 'filter_old_slug_redirect_url' ] );
+	}
+
+	/**
+	 * Flush rewrite rules.
+	 */
+	public function flush_rewrite_rules() {
+		flush_rewrite_rules( false );
+	}
+
+	/**
+	 * Add rewrite endpoint.
+	 */
+	public function add_rewrite_endpoint() {
+		if ( self::PERMALINK_STRUCTURE_REWRITE_ENDPOINT === AMP_Options_Manager::get_option( Option::PERMALINK_STRUCTURE ) ) {
+			$places = EP_ALL;
+		} else {
+			$places = EP_PERMALINK;
+		}
+		add_rewrite_endpoint( amp_get_slug(), $places );
+	}
+
+	/**
+	 * Add default option.
+	 *
+	 * @param array $defaults Default options.
+	 * @param array $options  Current options.
+	 * @return array Defaults.
+	 */
+	public function filter_default_options( $defaults, $options ) {
+		$value = self::PERMALINK_STRUCTURE_QUERY_VAR;
+
+		if (
+			isset( $options[ Option::VERSION ], $options[ Option::THEME_SUPPORT ], $options[ Option::READER_THEME ] )
+			&&
+			version_compare( $options[ Option::VERSION ], '2.1', '<' )
+		) {
+			if (
+				AMP_Theme_Support::READER_MODE_SLUG === $options[ Option::THEME_SUPPORT ]
+				&&
+				ReaderThemes::DEFAULT_READER_THEME === $options[ Option::READER_THEME ]
+			) {
+				$value = self::PERMALINK_STRUCTURE_LEGACY_READER;
+			} elseif ( AMP_Theme_Support::STANDARD_MODE_SLUG !== $options[ Option::THEME_SUPPORT ] ) {
+				$value = self::PERMALINK_STRUCTURE_LEGACY_TRANSITIONAL;
+			}
+		}
+
+		$defaults[ Option::PERMALINK_STRUCTURE ] = $value;
+
+		return $defaults;
+	}
+
+	/**
+	 * Sanitize options.
+	 *
+	 * @param array $options     Existing options with already-sanitized values for updating.
+	 * @param array $new_options Unsanitized options being submitted for updating.
+	 *
+	 * @return array Sanitized options.
+	 */
+	public function sanitize_options( $options, $new_options ) {
+		if (
+			isset( $new_options[ Option::PERMALINK_STRUCTURE ] )
+			&&
+			in_array( $new_options[ Option::PERMALINK_STRUCTURE ], self::PERMALINK_STRUCTURES, true )
+		) {
+			$options[ Option::PERMALINK_STRUCTURE ] = $new_options[ Option::PERMALINK_STRUCTURE ];
+		}
+		return $options;
+	}
+
+	/**
+	 * Get the current paired AMP permalink structure.
+	 *
+	 * @return string Paired AMP permalink structure.
+	 */
+	public function get_permalink_structure() {
+		$has_filters      = [
+			has_filter( 'amp_has_paired_endpoint' ),
+			has_filter( 'amp_add_paired_endpoint' ),
+			has_filter( 'amp_remove_paired_endpoint' ),
+		];
+		$has_filter_count = count( array_filter( $has_filters ) );
+		if ( 3 === $has_filter_count ) {
+			return self::PERMALINK_STRUCTURE_CUSTOM;
```

Comment:
As noted by @schlessera, we may want to switch this around so that the filters do not _force_ the custom URL structure, but rather that the filters can merely allow the custom structure to be among the selectable options.